radixdlt-engine/src/main/java/com/radixdlt/atom/TxAction.java

```diff
@@ -64,5 +64,216 @@
 
 package com.radixdlt.atom;
 
-/** Marker interface for RadixEngine actions. */
-public interface TxAction {}
+import static com.radixdlt.utils.Strings.asEmptyIfNull;
+import static java.util.Objects.requireNonNull;
+
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.identifiers.REAddr;
+import com.radixdlt.utils.Bytes;
+import com.radixdlt.utils.UInt256;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.function.LongFunction;
+
+/** RadixEngine actions. */
+public sealed interface TxAction {
+  record BurnToken(REAddr resourceAddr, REAddr fromAddr, UInt256 amount) implements TxAction {
+    public BurnToken {
+      ensureNonZeroAmount(amount);
+    }
+  }
+
+  record CreateFixedToken(
+      REAddr resourceAddr,
+      REAddr accountAddr,
+      String symbol,
+      String name,
+      String description,
+      String iconUrl,
+      String tokenUrl,
+      UInt256 supply)
+      implements TxAction {
+    public CreateFixedToken {
+      ensureResourceAddressIsHashedKey(resourceAddr);
+      ensureNonZeroSupply(supply);
+      requireAllNonNull(accountAddr, symbol, name, description, iconUrl, tokenUrl);
+    }
+  }
+
+  record CreateMutableToken(
+      REAddr resourceAddress,
+      String symbol,
+      String name,
+      String description,
+      String iconUrl,
+      String tokenUrl,
+      ECPublicKey owner)
+      implements TxAction {
+    public CreateMutableToken(
+        REAddr resourceAddress,
+        String symbol,
+        String name,
+        String description,
+        String iconUrl,
+        String tokenUrl,
+        ECPublicKey owner) {
+      this.resourceAddress = resourceAddress;
+      this.symbol = symbol.toLowerCase();
+      this.owner = owner;
+
+      this.name = asEmptyIfNull(name);
```

Comment:
I notice we validate this differently from the fixed token (because we're trying not to change behaviour) - do you think it might be worth explaining why or investigating whether we can safely do null checks on this instead of replacing with empty?

Reply:
I guess this somehow reflects business requirements, but I'm not aware about any relevant documents.